Create Openshift provider with Prometheus endpoint

zgalor opened the pull request:
Replaced `metrics: True` with default hawkular to generic endpoint terminology:
`monitoring: 'hawkular' / 'prometheus' / None`

endpoint params replaced accordingly:
`hawkular_hostname` => `monitoring_hostname`
`hawkular_port` => `monitoring_port`

Also added a creation example for prometheus in examples/add_openshift_provider.yml
and a test in tests/test_manageiq_provider.py

Temporary hack - No validation for prometheus endpoint authentication until the is supported by manageiq.

**Note**: this changes the module API and older playbooks with hawkular endpoint will not be supported from now on

zgalor committed:
Create openshift provider with prometheus endpoint

Added the ability to choose metrics_type: "prometheus" or "hawkular".
This is a new parameter, and it's value is "hawkular" by default for backward compatibility
If metric_type=prometheus then prometheus_hostname and prometheus_port must be defined

Also added a creation example for prometheus in examples/add_openshift_provider.yml
and a test in tests/test_manageiq_provider.py

Temporary hack - No validation for prometheus endpoint authentication until the is supported by manageiq.
This is also stated in the code in a todo

zgalor commented:
@cben @dkorn @moolitayer please review

cben commented on examples/add_openshift_provider.yml:
```diff
@@ -19,5 +19,25 @@
+      metrics: true
+      metrics_type: 'prometheus'
+      prometheus_hostname: 'prometheus01.example.com'
+      prometheus_port: '80'
```
So this needs 4 parameters, and could be 2 or 3.
`metrics: true` and `metrics_type: 'prometheus'` are IMO redundant.  I propose combining to `metrics: 'prometheus'`.

In theory, both could be infered from the fact you provided `prometheus_hostname:` and `prometheus_port:`, but IIRC we added separate `metrics: true` boolean to be able to ~~specify it but omit host/port which had a sensible default(?)~~

@dkorn @enoodle what do you think?

zgalor replied:
So how about:
metrics : 'hawkular' / 'prometheus' / none (by default)
Then of course we could have `metrics_port, metrics_hostname` regardless of the venue.

But this is not backwards compatible with old playbooks. do we want it to be?

cben replied:
Actually I don't see any trace of hawkular host/port being optional.

moolitayer replied:
I like best the option of having `hawkular_hostname`, `hawkular_port` OR `prometheus_hostname`, `prometheus_port`. Bad combinations should fail.
I would not touch `metrics: true` in this PR. 
I don't think we should be concerned with backward compatibility at this stage.

cben reviewed (commented):
Code LGTM. (aside from the interface metrics/metrics_type discussion)

zgalor committed:
Renaming for more generic usage

Renamed endpoint metrics to monitoring and hawkular/prometheus hostname + port united to monitoring_hostname/port
This is not backwards compatible with current playbooks but much more fitting and generic for future plans.

zgalor commented on library/manageiq_provider.py:
```diff
@@ -531,16 +551,18 @@ def main():
             ('provider_type', 'openshift-enterprise', ['provider_api_hostname', 'provider_api_port', 'provider_api_auth_token']),
-            ('metrics', True, ['hawkular_hostname', 'hawkular_port']),
+            ('monitoring', 'hawkular', ['monitoring_hostname', 'monitoring_port']),
+            ('monitoring', 'prometheus', ['monitoring_hostname', 'monitoring_port']),
```
I am not so happy with writing this validation twice - once for prometheus and once for hawkular, but I couldn't find a way in `required_if` to state "not None". Suggestions are welcome...

cben replied:
I don't think there is.
https://github.com/ansible/ansible/blob/b69c7f50d02dac3e24bf1927ebdc933475da93b7/lib/ansible/module_utils/basic.py#L1635 checks `== val`.
Are you unhappy with textual duplication (LGTM) or the fact we'll have to remember to extend it if we add another?

required_if is not magical.
If you want you can check the condition yourself and exit with error (like done below for  `['miq_url', 'miq_username', 'miq_password']`)...

zgalor replied:
I am unhappy with both, but then again it might be clearer to leave it like this. I the list or choices grows longer, we should refactor.

zgalor commented:
After a consult with @moolitayer I decided to rename the endpoint parameters to `monitoring`:
```
monitoring: prometheus / hawkular / None
monitoring_hostname
monitoring_port
```

This aligns with the new naming convention in manageiq, and more generic for future usage of logging etc.

@cben @yaacov @dkorn @moolitayer  please review

yaacov commented on examples/add_openshift_provider.yml:
```diff
@@ -2,17 +2,36 @@
+      miq_username: 'admin'
+      miq_password: '******'
+      miq_verify_ssl: false
+      provider_verify_ssl: false
```
I think it's nicer if the `provider_verify_ssl` will be together with the `provider_api_hostname` and `api_port`

zgalor replied:
I agree, but this is only an example, and also irrelevant to this PR

cben commented:
If "monitoring" term can include logging, will we at some point need 2 monitoring endpoints of different kinds?

zgalor commented:
@cben Probably not. More likely that each will reside in a different manager, @moolitayer keep me honest here.

cben commented on library/manageiq_provider.py:
```diff
@@ -531,16 +551,18 @@ def main():
-            hawkular_port=dict(required=False),
+            monitoring=dict(required=False, default=None,
+                       choices=['hawkular', 'prometheus', None]),
+            monitoring_hostname=dict(required=False),
```
Could add `type='str'`, also missing on many others. Not necessarily in this PR.

zgalor replied:
since we have "choices", this seems obsolete

cben commented on library/manageiq_provider.py:
```diff
@@ -577,8 +598,10 @@ def main():
+            if monitoring:
+                monitoring_hostname =  module.params['monitoring_hostname']
+                monitoring_port = module.params['monitoring_port']
+                endpoints.append(manageiq.generate_auth_key_config(monitoring, monitoring, monitoring_hostname, monitoring_port, token, provider_verify_ssl, provider_ca_path))
```
I'd pass these by name
```
generate_auth_key_config(
    role=monitoring,
    authtype=monitoring,
    ...
```

zgalor replied:
Very well.
So I guess this should be done for other calls, right?

cben replied:
Your call ;-).
Personally I find I can't read the current `generate_auth_key_config()` call without going to see its definition.  I suppose if you change this, should also change the 2 other endpoint-generating calls.

cben commented on tests/test_manageiq_provider.py:
```diff
@@ -86,7 +88,7 @@
 
 GET_RETURN_VALUES = {
-    'openshift_without_hawkular': {
+    'openshift_without_metrics': {
```
consider s/openshift_without_metrics/openshift_without_monitoring/g.

more importantly, there is `miq_ansible_module.params = {'metrics': True}` on L274

zgalor replied:
nice catch. thanks

zgalor replied:
I wonder how the test passed with the `metrics: true` param. I will fix the test

cben reviewed (commented):
Looks good.

zgalor committed:
Refactoring

explicit params for generate_auth_config
cleanup tests

zgalor commented:
I think this is ready now.
@cben @moolitayer please take a final look

cben commented:
@dkorn I expect you have new priorities :) I'm starting to merge without reviews from you; let us know if you want to stay in the loop.

cben commented:
@zgalor please edit PR description according to current interface.

zgalor commented:
done.

moolitayer commented:
:tada:
